cmd/snap/cmd_run.go

```diff
@@ -52,25 +60,52 @@ func init() {
 }
 
 func (x *cmdRun) Execute(args []string) error {
-	return snapRun(x.Positional.SnapApp, x.Command, args)
-}
-
-func getSnapInfo(snapName string) (*snap.Info, error) {
-	// we need to get the revision here because once we are inside
-	// the confinement the snapd API may be unavailable.
-	snaps, err := Client().List([]string{snapName})
-	if err != nil {
-		return nil, err
+	// Catch some invalid parameter combinations, provide helpful errors
+	if x.Hook != "" && x.Command != "" {
+		return fmt.Errorf("cannot use --hook and --command together")
+	}
+	if x.Revision != "" && x.Hook == "" {
+		return fmt.Errorf("-r can only be used with --hook")
+	}
+	if x.Hook != "" && len(args) > 0 {
+		return fmt.Errorf("too many arguments for hook %q: %s", x.Hook, strings.Join(args, " "))
 	}
-	if len(snaps) == 0 {
-		return nil, fmt.Errorf("cannot find snap %q", snapName)
+
+	// Now actually handle the dispatching
+	if x.Hook != "" {
+		return snapRunHook(x.Positional.SnapApp, x.Hook, x.Revision)
 	}
-	if len(snaps) > 1 {
-		return nil, fmt.Errorf("multiple snaps for %q: %d", snapName, len(snaps))
+
+	return snapRunApp(x.Positional.SnapApp, x.Command, args)
+}
+
+func getSnapInfo(snapName string, snapRevision string) (*snap.Info, error) {
+	var revision snap.Revision
+	if snapRevision != "" {
+		// User supplied a revision.
+		var err error
+		revision, err = snap.ParseRevision(snapRevision)
+		if err != nil {
+			return nil, fmt.Errorf("invalid revision: %q", snapRevision)
```

Comment:
ParseRevision should already be saying as much itself, so we must just return the err here if we don't want redundant messages.